collector/lib/ResourceSelector.cpp

```diff
@@ -0,0 +1,138 @@
+//
+// Created by Robby Cochran on 3/28/24.
+//
+
+#include "ResourceSelector.h"
+
+#include <regex>
+
+#include "Hash.h"
+
+namespace collector {
+
+UnorderedMap<std::string, std::regex> ResourceSelector::regexMap_;
+
+std::regex ResourceSelector::GetOrCompileRegex(const std::string& str) {
+  auto pair = regexMap_.find(str);
+  if (pair != regexMap_.end()) {
+    return pair->second;
+  } else {
+    std::regex pattern(str);
+    regexMap_.insert(std::make_pair(str, pattern));
```

Comment:
Asking out of ignorance here, what's the benefit of this versus something like this?
```cpp
regexMap_[str] = pattern;
```

Reply:
In this case there is no benefit to insert. The difference with insert, is that it will not overwrite the value if the key already exists. We don't have to worry about that in this case so assignment is better as it is more clear.